cpp/src/arrow/filesystem/filesystem.h

```diff
@@ -283,13 +283,17 @@ class ARROW_EXPORT FileSystem : public std::enable_shared_from_this<FileSystem>
   ///
   /// If the target already exists, existing data is truncated.
   virtual Result<std::shared_ptr<io::OutputStream>> OpenOutputStream(
-      const std::string& path) = 0;
+      const std::string& path,
+      const std::shared_ptr<const KeyValueMetadata>& metadata) = 0;
```

Comment:
Nit: the produced stream will always take ownership of `metadata`, so this shouldn't be a ref
```suggestion
      std::shared_ptr<const KeyValueMetadata> metadata) = 0;
```

Reply:
Not necessarily, the stream may not support it or it may convert the metadata to its own internal format.

Reply:
`KeyValueMetadata` is really just a glorified way of passing a collection of key-value pairs.